src/Microsoft.OData.Core/UriParser/Resolver/ODataUriResolver.cs

```diff
@@ -313,22 +342,20 @@ public virtual IEnumerable<IEdmOperationImport> ResolveOperationImports(IEdmMode
         /// <returns>The resolved key list.</returns>
         public virtual IEnumerable<KeyValuePair<string, object>> ResolveKeys(IEdmEntityType type, IList<string> positionalValues, Func<IEdmTypeReference, string, object> convertFunc)
         {
-            var keyProperties = type.Key().ToList();
-
             // Throw an error if key size from url doesn't match that from model.
             // Other derived ODataUriResolver intended for alternative key resolution, such as the built in AlternateKeysODataUriResolver,
             // should override this ResolveKeys method.
-            if (keyProperties.Count != positionalValues.Count)
+            if (type.Key().Count() != positionalValues.Count)
             {
                 throw ExceptionUtil.CreateBadRequestError(Strings.BadRequest_KeyCountMismatch(type.FullName()));
             }
 

```

Comment:
If we optimize for the success scenario I think we could avoid the Count().

1. first do the foreach loop while counting how many elements there are in Key().
2. after the foreach loop, check if the count of Key() != than positionalValues.Count. Throw exception if so.

Wouldn't this work?

Reply:
@joaocpaiva It might be counted as a breaking change. In that case, we will end up throwing the other exception before we could throw key count mismatch. 
I know this sounds like a bad request scenario anyway and people shouldn't care about it but we have had customers complain because their tests started failing after an upgrade.

Reply:
Yeah that makes sense. It is ok to leave as is then. Alternatively, we could also move the Syntax Error exception after the loop but it does not seem worth the trouble.